File: src/main/java/com/github/dylmeadows/eontimer/ui/settings/action/ActionSettingsModel.java
package com.github.dylmeadows.eontimer.ui.settings.action;

import com.github.dylmeadows.eontimer.handlers.ActionHandler;
import com.github.dylmeadows.eontimer.model.ActionMode;
import com.github.dylmeadows.eontimer.model.resources.SoundResource;
import com.github.dylmeadows.eontimer.reference.settings.ActionSettingsConstants;
import javafx.beans.property.IntegerProperty;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleIntegerProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.value.ObservableValue;
import javafx.scene.paint.Color;

/**
 * Settings related to timer action specific configuration. Binds to
 * {@link ActionSettingsView}.
 */
public class ActionSettingsModel implements ActionSettingsConstants {

    /**
     * See {@link com.github.dylmeadows.eontimer.ui.EonTimerController#onActionModeChange(ObservableValue, ActionMode, ActionMode)}
     * Binds to {@link ActionSettingsView#modeField}.
     */
    private final ObjectProperty<ActionMode> mode;

    /**
     * See {@link com.github.dylmeadows.eontimer.handlers.actions.VisualAction#activeFill}.
     * Binds to {@link ActionSettingsView#colorField}.
     */
    private final ObjectProperty<Color> color;

    /**
     * See {@link com.github.dylmeadows.eontimer.handlers.actions.SoundAction#sound}.
     * Binds to {@link ActionSettingsView#soundField}.
     */
    private final ObjectProperty<SoundResource> sound;

    /**
     * See {@link ActionHandler#interval}.
     * Binds to {@link ActionSettingsView#intervalField}.
     */
    private final IntegerProperty interval;

    /**
     * See {@link ActionHandler#count}.
     * Binds to {@link ActionSettingsView#countField}.
     */
    private final IntegerProperty count;

    /**
     * Initializes all properties to their default values.
     */
    public ActionSettingsModel() {
        mode = new SimpleObjectProperty<>(DEFAULT_MODE);
        color = new SimpleObjectProperty<>(DEFAULT_COLOR);
        sound = new SimpleObjectProperty<>(DEFAULT_SOUND);
        interval = new SimpleIntegerProperty(DEFAULT_INTERVAL);
        count = new SimpleIntegerProperty(DEFAULT_COUNT);
    }

    /**
     * @return see {@link #mode}
     */
    public ActionMode getMode() {
        return mode.get();
    }

    /**
     * @return see {@link #mode}
     */
    public ObjectProperty<ActionMode> modeProperty() {
        return mode;
    }

    /**
     * @param mode see {@link #mode}
     */
    public void setMode(ActionMode mode) {
        this.mode.set(mode);
    }

    /**
     * @return see {@link #color}
     */
    public Color getColor() {
        return color.get();
    }

    /**
     * @return see {@link #color}
     */
    public ObjectProperty<Color> colorProperty() {
        return color;
    }

    /**
     * @param color see {@link #color}
     */
    public void setColor(Color color) {
        this.color.set(color);
    }

    /**
     * @return see {@link #sound}
     */
    public SoundResource getSound() {
        return sound.get();
    }

    /**
     * @return see {@link #sound}
     */
    public ObjectProperty<SoundResource> soundProperty() {
        return sound;
    }

    /**
     * @param sound see {@link #sound}
     */
    public void setSound(SoundResource sound) {
        this.sound.set(sound);
    }

    /**
     * @return see {@link #interval}
     */
    public int getInterval() {
        return interval.get();
    }

    /**
     * @return see {@link #interval}
     */
    public IntegerProperty intervalProperty() {
        return interval;
    }

    /**
     * @param interval see {@link #interval}
     */
    public void setInterval(int interval) {
        this.interval.set(interval);
    }

    /**
     * @return see {@link #count}
     */
    public int getCount() {
        return count.get();
    }

    /**
     * @return see {@link #count}
     */
    public IntegerProperty countProperty() {
        return count;
    }

    /**
     * @param count see {@link #count}
     */
    public void setCount(int count) {
        this.count.set(count);
    }
}

File: src/models/settings/ActionSettingsModel.cpp
//
// Created by Dylan Meadows on 2020-03-15.
//

#include "ActionSettingsModel.h"

namespace model::settings {
    namespace ActionSettingsFields {
        const char *GROUP = "action";

        const char *MODE = "mode";
        const char *SOUND = "sound";
        const char *COLOR = "color";
        const char *INTERVAL = "interval";
        const char *COUNT = "count";

        namespace Defaults {
            const int MODE = 0;
            const int SOUND = 0;
            const QColor COLOR = Qt::blue;
            const unsigned int INTERVAL = 500;
            const unsigned int COUNT = 6;
        }  // namespace Defaults
    }      // namespace ActionSettingsFields

    ActionSettingsModel::ActionSettingsModel(QSettings *settings) : QObject(nullptr) {
        settings->beginGroup(ActionSettingsFields::GROUP);
        mode = model::actionMode(
            settings->value(ActionSettingsFields::MODE, ActionSettingsFields::Defaults::MODE).toUInt());
        sound =
            model::sound(settings->value(ActionSettingsFields::SOUND, ActionSettingsFields::Defaults::SOUND).toUInt());
        color = settings->value(ActionSettingsFields::COLOR, ActionSettingsFields::Defaults::COLOR).value<QColor>();
        interval = settings->value(ActionSettingsFields::INTERVAL, ActionSettingsFields::Defaults::INTERVAL).toUInt();
        count = settings->value(ActionSettingsFields::COUNT, ActionSettingsFields::Defaults::COUNT).toUInt();
        settings->endGroup();
    }

    void ActionSettingsModel::sync(QSettings *settings) const {
        settings->beginGroup(ActionSettingsFields::GROUP);
        settings->setValue(ActionSettingsFields::MODE, model::indexOf(mode));
        settings->setValue(ActionSettingsFields::SOUND, model::indexOf(sound));
        settings->setValue(ActionSettingsFields::INTERVAL, interval);
        settings->setValue(ActionSettingsFields::COUNT, count);
        settings->setValue(ActionSettingsFields::COLOR, color);
        settings->endGroup();
    }

    model::ActionMode ActionSettingsModel::getMode() const { return mode; }

    void ActionSettingsModel::setMode(model::ActionMode mode) { this->mode = mode; }

    model::Sound ActionSettingsModel::getSound() const { return sound; }

    void ActionSettingsModel::setSound(const model::Sound sound) { this->sound = sound; }

    const QColor &ActionSettingsModel::getColor() const { return color; }

    void ActionSettingsModel::setColor(const QColor &color) {
        if (this->color != color) {
            this->color = color;
            emit colorChanged(color);
        }
    }

    unsigned int ActionSettingsModel::getInterval() const { return interval; }

    void ActionSettingsModel::setInterval(const unsigned int interval) { this->interval = interval; }

    unsigned int ActionSettingsModel::getCount() const { return count; }

    void ActionSettingsModel::setCount(const unsigned int count) { this->count = count; }
}  // namespace model::settings
